doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the logging should be
+forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It shall also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal an non-recoverable errors which essentially leads to a termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens but not considered as severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independent but in
+combination with the error handler.
+
+## Design
+
+### Considerations
+
+1. Similar to other iceoryx constructs, heap usage, exceptions and recursive
```

Comment:
Well, it depends. Our MemPools are also dynamic memory but handled by our self.

Reply:
Sure, we do not want to use it though even if we manage it ourselves because we might run out at runtime. That is why I prefer a more general statement here. Allocating before logging begins is of course fine (and not dynamic).